lib/ecto/query/builder.ex

```diff
@@ -1104,6 +1103,27 @@ defmodule Ecto.Query.Builder do
           "`#{Macro.to_string(other)}`"
       )
 
+  @doc """
+  Checks if the field is an atom or string at compilation time or
+  delegate the check to runtime for interpolation.
+  """
+  def quoted_atom_or_string!({:^, _, [expr]}, used_ref),
+    do: quote(do: Ecto.Query.Builder.atom_or_string!(unquote(expr), unquote(used_ref)))
+
+  def quoted_atom_or_string!(atom, _used_ref) when is_atom(atom),
+    do: atom
+
+  def quoted_atom_or_string!(string, _used_ref) when is_binary(string),
+    do: string
+
+  def quoted_atom_or_string!(other, used_ref),
+    do:
+      error!(
+        "expected literal atom or string or interpolated value in #{used_ref}, got: " <>
```

Comment:
```suggestion
        "expected literal atom/string or interpolated value in #{used_ref}, got: " <>
```
Makes it more obvious to me to understand you mean `literal*(atom+string) + interpolated`.